### PR TITLE
Add chainId support to TokenRatesController

### PR DESCRIPTION
**Pull request by wachunei**

### CODE REVIEW

File: src/assets/TokenRatesController.ts
Comment (wachunei) [inline src/assets/TokenRatesController.ts:98]: Should this be wrapped in a `try catch` and on the `catch` block have the same case `this.update({ chainSlugIdentifier: '' });`?

File: src/assets/TokenRatesController.ts
Comment (wachunei) [inline src/assets/TokenRatesController.ts:180]: When `chainId` changes, run update if not disabled.

File: src/assets/TokenRatesController.ts
Comment (wachunei) [inline src/assets/TokenRatesController.ts:217]: If this was called directly, nothing was stopping it from being run. What is the intended behavior originally: should disable just disable tokenList/chainId changes?

File: src/assets/TokenRatesController.ts
Comment (wachunei) [inline src/assets/TokenRatesController.ts:234]: Here's the main question. 

Since now there is an extra chainId parameter, we need to hit two endpoints
1. Get the chain slug given the chain id
2. Use the chain slug to recover the prices, like it was done before

**Should we "remember" the slug?** There are two three cases:
1. The request fails
2. The request is successful but the requested chain id is not listed on the api
3. The request is successful and we have a slug we can use

If we don't have the slug stored, this will always make two request when fetching prices.
If we store the slug, we must be able to define when to invalidate it in case the API changes, for example: it supports a new chain which we stored as unsupported

**Comment by wachunei** [comment]
Another issue is that `updateExchangeRates` will run twice: when chainId changes and when the token list changes. 

This could be avoided if assets had the chain_id as property, which would also work when listing assets from different chains in the same view.

File: src/assets/TokenRatesController.ts
Comment (adonesky1) [reply src/assets/TokenRatesController.ts:234]: Maybe we store the slug when we are able to get it and then wrap the `fetchExchangeRate` in a try/catch in case the slug becomes invalid and attempt to refetch it if that's the case?

Comment (wachunei) [reply src/assets/TokenRatesController.ts:234]: How would we manage the case the slug does not exist and later it does? Example: We would store it as `null`, so we don't even attempt to fetch the rates, but if the slug is supported in the future, how would we invalidate that `null`?

Comment (adonesky1) [reply src/assets/TokenRatesController.ts:234]: Maybe I'm not understanding but if the chainSlugIdentifier is null we try will try to fetch it again as the code stands? Its not ideal because we are not able to cut out that extra network request by storing it as unsupported but I can't think of a way around that right now - unless they only update the api when they change the endpoint (v3 -> v4)?

Comment (wachunei) [reply src/assets/TokenRatesController.ts:234]: What we are doing on `swaps-controller` to hold values for a certain amount of time is to store pairs `(value, timestamp)` and compare `timestamp`, which represents the last time `value` was fetched, to a time threshold.

This way if `value` is `null` but `Date.now() - timestamp` is less than a threshold, we consider the network is not supported, otherwise we attempt to fetch the slug and update the pair `(value, timestamp = Date.now())`.

The same works when `value` is `a-network-slug` that might become unsupported, but we check only after we are over the time threshold.

Does it make sense?

Comment (adonesky1) [reply src/assets/TokenRatesController.ts:234]: Yes that makes sense. So we're weighing the trade-offs of optimizing for fewer network requests and, if we choose to go that route, possibly not fetching exchange rates when they are available until the unsupported flag expires according to whatever time threshold we set. I don't have a strong intuition of which way to go with this. What do you think @Gudahtt?

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [inline src/assets/TokenRatesController.ts:93]: This function seems.... odd. It just combines the two parameters? Seems to have limited utility at first glance.

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [inline src/assets/TokenRatesController.ts:156]: We should not have a hard-coded default network - we should find out for sure what the initial network is. This could be disastrous, if this controller thinks the user is on a different network until he first switch.

Comment (Gudahtt) [reply src/assets/TokenRatesController.ts:156]: Kinda just as bad to do this with the native currency as well to be honest, though that's a pre-existing issue.

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [inline src/assets/TokenRatesController.ts:195]: We've been trying to move away from using setters for this kinda thing, and instead preferring methods. If this was a method (like `setChainId`) it's clear that setting it has a side-effect, and the caller can look to the method definition to see what it does. It might even have a nice JSDoc comment (which gets shown inline via intellisense on some editors).

But with a setter like this... it's ambiguous to the caller what effect setting this has. It might have a setter, or it might be referenced anywhere in this controller or in other modules, or both.

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [inline src/assets/TokenRatesController.ts:258]: This branch seems like it should be impossible to reach :thinking: Since `supportedChains[chainId]` is set immediately after switching networks.

**Review by Gudahtt — COMMENTED** [review]
I left some initial comments here and on Notion. If we decide to pursue caching the API response, some of these comments might end up being obsolete though.

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:93]: Yeah, it actually merges a nested property of state because `BaseController.update` is shallow merge, would override other `chainIds` values.

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [reply src/assets/TokenRatesController.ts:297]: I don't think the `disabled` property was used in this controller :thinking:

Like many aspects of the BaseController, some controllers are written to use it and some ignore it, or break when you try to use it. The config is similar - it's built to be changeable at runtime, but many controllers are written assuming that never happens, and they break if you try it.

It doesn't hurt to add a check for `disabled` here but I'm not sure it helps either, unless I'm missing something. This awkwardness with the API is one of the things BaseControllerV2 is meant to address.

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:156]: This was done in order to provide backwards compatibility, since the controller was implicitly always hard coded to `chainId: 1` (because `ethereum` was the hardcoded endpoint).

Will look into how we should handle this properly then.

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:195]: Is there an example of the right way to do this?
I've been using it as a side-effect handler of `controller.config({...})`

Comment (Gudahtt) [reply src/assets/TokenRatesController.ts:195]: Ah, right, I forgot `configure` worked that way. Ah well. Maybe this is fine until switching to the new BaseControllerV2 API then. Not a huge deal.

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [reply src/assets/TokenRatesController.ts:156]: I've handled this elsewhere by passing in the current network as a constructor parameter, or passing in a "getNetworkState" function which can be used to check the initial state.

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:258]: The race condition I was thinking of is:
Given this state:
- `chainCache.slug` is `null`
- `tokens` is a list of tokens

Since we have now two listeners (on assets change and on network change), could it be the case this method executes when the assets change (because of a network change) instead of the network change itself?

There is an open question on the technical proposal about this

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:156]: Cool, will look it up!

File: src/assets/TokenRatesController.ts
Comment (Gudahtt) [reply src/assets/TokenRatesController.ts:258]: The thing is, if the assets change before the chain ID does, this would still be referencing the old chain ID. I don't believe it's possible for this to have a new chain ID without the setter for chainId running first.

Comment (wachunei) [reply src/assets/TokenRatesController.ts:258]: I see it now, this check is unnecessary 👍

File: src/assets/TokenRatesController.ts
Comment (adonesky1) [inline src/assets/TokenRatesController.ts:325]: [nit] @wachunei I know you didn't name this const but it seems misnamed to me? It's a list of tokens that are each one side of a pair with the native currency right?

Comment (wachunei) [reply src/assets/TokenRatesController.ts:325]: No, this is just a concatenation of token addresses joined by a comma, from something like `[{ address: '0x1...', ...restOfTokenObject1 }, { address: '0x2...', ...restOfTokenObject2 }]` into `"0x1...,0x2..."`, which is what CoinGecko takes in as query param for the API call.

Comment (wachunei) [reply src/assets/TokenRatesController.ts:325]: I agree is not a great name! But you know, naming stuff is hard hahahaha

**Comment by adonesky1** [comment]
@wachunei have you tested this at all wired up in the extension?

**Comment by wachunei** [comment]
> @wachunei have you tested this at all wired up in the extension?

No, I could test it with the mobile app, have not done it yet tho.

**Comment by adonesky1** [comment]
> No, I could test it with the mobile app, have not done it yet tho.

Ok I'm going to give it a go (on the extension) if there are no objections.

**Comment by wachunei** [comment]
That would be great! Thank you @adonesky1

**Comment by adonesky1** [comment]
@Gudahtt, @brad-decker  is there any roadmap for migrating the extension towards use of the AssetsController? Or, if not, in order to integrate the TokenRatesController in the extension will we need to modify this controller to expect alternative listeners that correspond with where/how token data is in the extension?

**Comment by Gudahtt** [comment]
It is planned, but only in our imaginations at this point. No concrete plan yet.

**Comment by wachunei** [comment]
@adonesky1 I tested in on the app and it is working, opened a draft PR of the integration here https://github.com/MetaMask/metamask-mobile/pull/2863

Screen Shot 2021-07-01 at 10 41 13Screen Shot 2021-07-01 at 10 40 55Screen Shot 2021-07-01 at 10 40 27

File: src/assets/TokenRatesController.ts
Comment (brad-decker) [inline src/assets/TokenRatesController.ts:91]: [Q] Is this a common pattern in the controller code base, having the UPPER_SNAKE_CASE (which I read as a constant) have methods on it? Non blocking just curious

**Review by brad-decker — DISMISSED** [review]
Looks good to me!One non blocking Q

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:91]: Great question, actually I don't know. It has landed to this point because it started as a constant plain object but then it mutated to an object with methods for convenience. Happy to adopt convention/pattern before moving on.

Comment (brad-decker) [reply src/assets/TokenRatesController.ts:91]: Maybe just `CoinGeckoApi` or some such for now. I could see it living in its own `platforms` file when there are more than one platform we use for this. For now its fine here. Its also fine to let it in like this.

Comment (wachunei) [reply src/assets/TokenRatesController.ts]: Changed on [`1aa1bfb` (#476)](https://github.com/MetaMask/controllers/pull/476/commits/1aa1bfbad2873634dc83ae2ca929f8757e60cc34)

**Review by brad-decker — APPROVED** [review]

**Comment by adonesky1** [comment]
> Ok I'm going to give it a go (on the extension) if there are no objections.

@wachunei sorry I never followed up on this. This work is blocked by a need to migrating the extension to use of the AssetController. I began to plan this work and then got pulled in another direction. This obviously shouldn't be a blocker though. I will probably try to pick that work up again soon after this has merged.

**Comment by wachunei** [comment]
@adonesky1 No worries. 

This is also in a weird state because of develop moving on to `v11`. Do we maintain minor versions of different major versions in parallel?
- If I revert [`e4bc0d1`](https://github.com/MetaMask/controllers/pull/476/commits/e4bc0d183f3ee43e20f98513e365fd59014413fa) on this branch, this is `10.x` compatible (our current mobile target version) and this could be released as `10.3`
- If I merge as is, this would be `11` compatible and won't work with current mobile code, so this is blocked until we upgrade to `11`

Thoughts?

**Comment by Gudahtt** [comment]
Generally I'd recommend always targeting the main branch with all bug fixes/features/etc. and separately backporting if necessary. If you're going to be backporting, you'll typically want the change in both places anyway, and it's just easier to land it on `main` first.

There are some basic instructions for how to prepare a backport release in the README. Basically you'd create a `10.x` branch off of `v10.1` (the most recent non-deprecated v10 release), create a release branch based there, cherry-pick the change you want to backport, and do the usual release bump and changelog update. Whether you want to bother with backporting is up to you/the mobile team. It's a tradeoff between taking the extra effort to backport v.s. waiting for v11 to land. It's probably better to land v11 and save yourself the effort unless time constraints make that difficult.

**Comment by wachunei** [comment]
Thanks @Gudahtt, I'll merge this and ask the mobile team about having this included in v11 or backporting.

**Comment by wachunei** [comment]
Just a reminder this would be a major version bump since `TokenRatesController` constructor now expects a `onNetworkStateChange` listener.

File: src/assets/TokenRatesController.ts
Comment (wachunei) [reply src/assets/TokenRatesController.ts:325]: `pairs` should be `addresses`, sorry for necroposting

**Pull request merged**
